tools/ExampleExtractor/ExtractorAndTesterUsersGuide.md

```diff
@@ -0,0 +1,483 @@
+
+If multiple *annotation_directive*s contain the same name, the right-most directive prevails.
+
+The ordering of annotation directives is not significant.
+
+The C# spec contains two kinds of source-code examples: those within an \*Example\* and those not, as follows:
+
+````
+> *Example*:
+>
+> <!-- Example: { … } -->
+> ```csharp
+> // source code goes here
+> ```
+>
+> *end example*
+````
+
+In this case, the *example_annotation* is preceded by `> ` to match the example’s formatting. However, in the following case, it is not, as no \*Example\* is present:
+
+````
+<!-- Example: { … } -->
+```csharp
+// source code goes here
+```
+````
+
+### Templates
+
+An annotation’s *template* is used to select certain compiler options, and to provide supporting machinery, as needed. This *annotation_directive* is required.
+
+```ANTLR
+template
+    : 'template' ':' template_name
+    ;
+template_name
+    : '"standalone-console"'  // actually, a JSON_string_value with this content
+    | '"standalone-lib"'      // actually, a JSON_string_value with this content
+    | '"code-in-main"'        // actually, a JSON_string_value with this content
+    ;
+```
+
+The template `standalone-console` indicates that the example is an application. For example:
+
+````
+> <!-- Example: {template:"standalone-console", …} -->
+> ```csharp
+> class Hello
+> {
+>     static void Main()
+>     {
+>         …
+>     }
+> }
+> ```
+````
+
+The template `standalone-lib` indicates that the example is a class library. For example:
+
+````
+> <!-- Example: {template:"standalone-lib", … } -->
+> ```csharp
+> class Class1
+> {
+>     static void Test(bool status)
+>     {
+>         …
+>     }
+> }
+> ```
+````
+
+The template `code-in-main` indicates that the example needs to be wrapped inside an entry-point method inside a class. For example, the example:
+
+````
+> <!-- Example: {template:"code-in-main", … } -->
+> ```csharp
+> int[][] pascals = 
+> {
+>     new int[] {1},
+>     new int[] {1, 1},
+>     new int[] {1, 2, 1},
+>     new int[] {1, 3, 3, 1}
+> };
+> ```
+````
+
+gets transformed into the following application:
+
+````
+class Program
+{
+    static void Main()
+    {
+        int[][] pascals = 
+{
+    new int[] {1},
+    new int[] {1, 1},
+    new int[] {1, 2, 1},
+    new int[] {1, 3, 3, 1}
+};
+    }
+}
+````
+
+### Example Names
+
+An annotation’s *name* is the name of the resulting test file. This *annotation_directive* is required.
+
+```ANTLR
+name
+    : 'name' ':' test_filename
+    ;
+test_filename
+    : JSON_string_value
+    ;
+```
+
+The ExampleExtractor tool processes all md files in a given input folder, and writes out all corresponding test files to a given output folder. As such, ***test_filenames*** **must be unique across the whole C# spec.** (If they are not, each subsequently generated same-named file will overwrite the previous one!)
+
+In the following example, the class-library source-code is written to file “NestedClassDependency”:
```

Comment:
```suggestion
In the following example, the class-library source-code is written to directory “NestedClassDependency”:
```

(The filename is part of the template; it's only the directory name that's changed by the name. If I've remembered correctly, at least...)

Reply:
OK, fixed.